url78117c6033_BrianMoths_LittleManManager_tgz-pJ8-littlemangame_LittleManGameJ8/src/littlemangame/notebookdeveloper/notebooktester/ExpectedProgramBehavior.java

```diff
@@ -0,0 +1,146 @@
+import littlemangame.notebookdeveloper.notebooktester.inputoutputevents.InputOutputEventType;
+import littlemangame.notebookdeveloper.notebooktester.inputoutputevents.OutputEvent;
+import littlemangame.word.Word;
+
+/**
+ *
+ * @author brian
+ */
+public class ExpectedProgramBehavior {
+
+    private final Queue<InputOutputEventType> inputOutputEventTypes;
+    private final Queue<Word> inputWords;
+    private final Queue<WordPredicate> outputWordPredicates;
+    private final StringBuilder errorStringBuilder;
+
+    public ExpectedProgramBehavior() {
+        inputOutputEventTypes = new ArrayDeque<>();
+        inputWords = new ArrayDeque<>();
+        outputWordPredicates = new ArrayDeque<>();
+        errorStringBuilder = new StringBuilder();
+    }
+
+    public ExpectedProgramBehavior(ExpectedProgramBehavior expectedProgramBehavior) {
+        inputOutputEventTypes = new ArrayDeque<>(expectedProgramBehavior.inputOutputEventTypes);
+        inputWords = new ArrayDeque<>(expectedProgramBehavior.inputWords);
+        outputWordPredicates = new ArrayDeque<>(expectedProgramBehavior.outputWordPredicates);
+        errorStringBuilder = new StringBuilder(expectedProgramBehavior.errorStringBuilder);
+    }
+
+    public void copy(ExpectedProgramBehavior expectedProgramBehavior) {
+        inputOutputEventTypes.clear();
+        inputOutputEventTypes.addAll(expectedProgramBehavior.inputOutputEventTypes);
+        inputWords.clear();
+        inputWords.addAll(expectedProgramBehavior.inputWords);
+        outputWordPredicates.clear();
+        outputWordPredicates.addAll(expectedProgramBehavior.outputWordPredicates);
+        errorStringBuilder.setLength(0);
+        errorStringBuilder.append(expectedProgramBehavior.errorStringBuilder);
+    }
+
+    public void addInputEvent(Word inputWord) {
+        inputOutputEventTypes.add(InputOutputEventType.INPUT);
+        inputWords.add(inputWord);
+    }
+
+    public void addOutputEvent(Word outputWord) {
+        addOutputEvent(WordPredicate.makeEqualsWordPredicate(outputWord));
+    }
+
+    public void addOutputEvent(WordPredicate wordPredicate) {
+        inputOutputEventTypes.add(InputOutputEventType.OUTPUT);
+        outputWordPredicates.add(wordPredicate);
+    }
+
+    public void addHaltEvent() {
+        inputOutputEventTypes.add(InputOutputEventType.HALT);
+    }
+
+    public boolean testInputEvent(InputEvent actualInputEvent) {
+        final InputOutputEventType expectedInputOutputEventType = inputOutputEventTypes.poll();
+        boolean isCorrectSoFar;
+        if (expectedInputOutputEventType != actualInputEvent.getEventType()) {
+            errorStringBuilder.append(makeErrorLineStringBuilder(expectedInputOutputEventType, actualInputEvent));
+            isCorrectSoFar = false;
+        } else {
+            errorStringBuilder.append(getActualActionStringBuilder(actualInputEvent));
+            isCorrectSoFar = true;
+        }
+        return isCorrectSoFar;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 72:81, 89:98.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.